community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaKey.java

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.index.schema;
+
+import java.util.Arrays;
+
+import org.neo4j.index.internal.gbptree.GBPTree;
+import org.neo4j.string.UTF8;
+import org.neo4j.values.storable.TextValue;
+import org.neo4j.values.storable.Value;
+import org.neo4j.values.storable.ValueWriter;
+import org.neo4j.values.storable.Values;
+
+import static java.lang.String.format;
+
+import static org.neo4j.values.storable.UTF8StringValue.codePointByteArrayCompare;
+
+/**
+ * Includes value and entity id (to be able to handle non-unique values). A value can be any {@link String},
+ * or rather any string that {@link GBPTree} can handle.
+ */
+class StringSchemaKey extends ValueWriter.Adapter<RuntimeException> implements NativeSchemaKey
+{
+    static final int ENTITY_ID_SIZE = Long.BYTES;
+
+    private long entityId;
+    private boolean entityIdIsSpecialTieBreaker;
+
+    // TODO something better or?
+    // TODO this is UTF-8 bytes for now
+    byte[] bytes;
+
+    @Override
+    public void setEntityIdIsSpecialTieBreaker( boolean entityIdIsSpecialTieBreaker )
+    {
+        this.entityIdIsSpecialTieBreaker = entityIdIsSpecialTieBreaker;
+    }
+
+    @Override
+    public boolean getEntityIdIsSpecialTieBreaker()
+    {
+        return entityIdIsSpecialTieBreaker;
+    }
+
+    int size()
+    {
+        return ENTITY_ID_SIZE + bytes.length;
+    }
+
+    @Override
+    public long getEntityId()
+    {
+        return entityId;
+    }
+
+    @Override
+    public void setEntityId( long entityId )
+    {
+        this.entityId = entityId;
+    }
+
+    @Override
+    public void from( long entityId, Value... values )
+    {
+        this.entityId = entityId;
+        entityIdIsSpecialTieBreaker = false;
+        assertValidValue( values ).writeTo( this );
+    }
+
+    private TextValue assertValidValue( Value... values )
+    {
+        if ( values.length > 1 )
```

Comment:
this more or less duplicate capability checks, maybe use that instead?

Reply:
Yes, the validation of "what goes where" happens all over the place right now. Created a card for cleaning this up. Ok if we leave as future work?

Reply:
oke